ynr/apps/sopn_parsing/helpers/extract_pages.py

```diff
@@ -31,7 +37,8 @@ def extract_pages_for_single_document(document):
     if not doc_file:
         return
 
-    if all_documents_with_source.count() == 1:
+    # if manual upload dont check for shared sources as this might be the first
+    if not manual_upload and all_documents_with_source.count() == 1:
```

Comment:
So this is going to be a very rare case, but consider:

1. The user manually uploads a milti-ballot SOPN
2. The ballots in the SOPN contain `Foo North ward` and `Foo ward`
3. This is a upload against `Foo ward`
4. `Foo North ward` is first in the PDF
5. I think this ballot will get matched to `Foo North ward` not `Foo ward`

This is because the parser sorts the possible ballots by the post label length, and then assumes the first match is right. 

I don't really know how to fix this though! Maybe an interface for manually fixing the matched pages would work, and might be a good way to fix misc problems with the parsing later on.

For now, let's leave the new feature, but if you can think of some simple way to mitigate the above then please have a crack at it!

Reply:
One other thing to consider: If this is a by-election (as is likely for manual uploads - we get most of them from the weekly by-elections) then two things are likely to be true

1. This is a single page SOPN - we can assume a one to one between the page and ballot
2. If not a single page, this is very likely to be for a single by-election. I don't have the numbers, but it's quite unlikely to have more than one by-election for a council in the same week, and if so I suspect there will be a PDF per ballot. 

This is worth researching (on past ballots), but it might be worth considering looking at `.by.` in the ballot paper ID and special casing that.

Reply:
Interesting, I knew there would be some edge case that I would not be familiar with! I will have a look at the page parsing, as it feels like something that should be fixed there. 

Also that's useful to know about by-elections. I'll do a bit of research on past ballots like suggested, but yes might be worth special casing a by-election and returning `all` for those.

Also I'm happy to remove the `manual_upload` force page extraction entirely if you think it is going to be more problematic than it is worth - it was something of an afterthought when I noticed that using a multi-page SOPN parsing failed. But if in reality its solving a problem that doesn't exist (because that's not how SOPN's for multiple wards are uploaded in practice) then it could be better to remove?

Reply:
Here's another idea that might solve other problems - because the problem I outlined might happen when we import from the spreadsheet too:

When matching pages, we _always_ look at all ballots in the election (sibling ballots).

We then match each page to each ballot, but only store the pages matched against the `OfficialDocument` that's been uploaded.

This might end up as quite a big refactor, but it would mean that we don't need to worry about matching partial SOPNs at all.

Reply:
As discussed on the call - will leave this for further work later. For now I have taken your comments and added them as an issue so they don't get lost and forgotten, but feel free to update as you like https://github.com/DemocracyClub/yournextrepresentative/issues/1385